File: backend/src/main/java/com/backend/domain/repository/util/LanguageUtils.java
package com.backend.domain.repository.util;

import com.backend.domain.repository.entity.Language;

public class LanguageUtils {
    public static Language fromGitHubName(String name) {
        if(name == null || name.isBlank()) return Language.OTHER;

        String normalized = name.trim().toLowerCase()
                .replace("+", "P")
                .replace("#", "sharp")
                .replace("-", "")
                .replace(" ", "");

        return switch (normalized) {
            case "java" -> Language.JAVA;
            case "python" -> Language.PYTHON;
            case "c" -> Language.C;
            case "cpp" -> Language.CPP;
            case "csharp" -> Language.CSHARP;
            case "go" -> Language.GO;
            case "rust" -> Language.RUST;
            case "kotlin" -> Language.KOTLIN;
            case "swift" -> Language.SWIFT;
            case "ruby" -> Language.RUBY;
            case "php" -> Language.PHP;
            case "objectivc", "objectivec" -> Language.OBJECTIVEC;
            case "javascript" -> Language.JAVASCRIPT;
            case "typescript" -> Language.TYPESCRIPT;
            case "html" -> Language.HTML;
            case "css" -> Language.CSS;
            case "scss", "sass" -> Language.SCSS;
            case "vue" -> Language.VUE;
            case "r" -> Language.R;
            case "sql" -> Language.SQL;
            case "shell", "bash", "sh" -> Language.SHELL;
            case "yaml", "yml" -> Language.YAML;
            case "json" -> Language.JSON;
            case "perl" -> Language.PERL;
            case "dart" -> Language.DART;
            case "haskell" -> Language.HASKELL;

            default -> Language.OTHER;
        };
    }
}

File: backend/src/main/java/com/backend/domain/repository/util/LanguageUtils.kt
package com.backend.domain.repository.util

import com.backend.domain.repository.entity.Language

object LanguageUtils {

    fun fromGitHubName(name: String?): Language {
        val normalized = name
            ?.trim()
            ?.lowercase()
            ?.replace("+", "P")
            ?.replace("#", "sharp")
            ?.replace("-", "")
            ?.replace(" ", "")
            ?: return Language.OTHER

        return when (normalized) {
            "java" -> Language.JAVA
            "python" -> Language.PYTHON
            "c" -> Language.C
            "cpp" -> Language.CPP
            "csharp" -> Language.CSHARP
            "go" -> Language.GO
            "rust" -> Language.RUST
            "kotlin" -> Language.KOTLIN
            "swift" -> Language.SWIFT
            "ruby" -> Language.RUBY
            "php" -> Language.PHP
            "objectivc", "objectivec" -> Language.OBJECTIVEC
            "javascript" -> Language.JAVASCRIPT
            "typescript" -> Language.TYPESCRIPT
            "html" -> Language.HTML
            "css" -> Language.CSS
            "scss", "sass" -> Language.SCSS
            "vue" -> Language.VUE
            "r" -> Language.R
            "sql" -> Language.SQL
            "shell", "bash", "sh" -> Language.SHELL
            "yaml", "yml" -> Language.YAML
            "json" -> Language.JSON
            "perl" -> Language.PERL
            "dart" -> Language.DART
            "haskell" -> Language.HASKELL

            else -> Language.OTHER
        }
    }
}
